README.md
# Helix Content Bus

> Serverless action that pushes helix-content-proxy into a storage service

**Note**: The content-bus action is no longer used. its functionality is now embedded in helix-admin.

## Status
[![codecov](https://img.shields.io/codecov/c/github/adobe/helix-content-bus.svg)](https://codecov.io/gh/adobe/helix-content-bus)
[![CircleCI](https://img.shields.io/circleci/project/github/adobe/helix-content-bus.svg)](https://circleci.com/gh/adobe/helix-content-bus)
[![GitHub license](https://img.shields.io/github/license/adobe/helix-content-bus.svg)](https://github.com/adobe/helix-content-bus/blob/main/LICENSE.txt)
[![GitHub issues](https://img.shields.io/github/issues/adobe/helix-content-bus.svg)](https://github.com/adobe/helix-content-bus/issues)
[![LGTM Code Quality Grade: JavaScript](https://img.shields.io/lgtm/grade/javascript/g/adobe/helix-content-bus.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/adobe/helix-content-bus)
[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)

## Installation

## Usage

The following parameters should be passed when invoking the action:
- `owner`: GitHub repository owner
- `repo`: GitHub repository name
- `ref`: GitHub repository reference or branch
- `path`: path to the document to be fetched from helix-content-proxy
- `prefix`: prefix to add to the S3 address, defaults to `live`

Note: the first three parameters also determine the location where the `fstab.yaml` configuration file is downloaded from.
The service requires that file to find a matching mount point, which also determines the bucket name used in S3.

The following environment variables are optional:
- `AWS_S3_REGION`: AWS region
- `AWS_S3_ACCESS_KEY_ID`: AWS access key associated with an IAM user or role
- `AWS_S3_SECRET_ACCESS_KEY`: Specifies the secret key associated with the access key

If they're not specified, the function operates with whatever role it was deployed with.

## Development

### Deploying Helix Content Bus

All commits to main that pass the testing will be deployed automatically. All commits to branches that will pass the testing will get commited as `/helix-services/content-bus@ci<num>` and tagged with the CI build number.


CHANGELOG.md
# [2.0.0](https://github.com/adobe/helix-content-bus/compare/v1.3.18...v2.0.0) (2021-08-04)


### Bug Fixes

* use single bucket for content-bus ([be6e183](https://github.com/adobe/helix-content-bus/commit/be6e1830a4c39a0af551d232a617606261c411a2)), closes [#88](https://github.com/adobe/helix-content-bus/issues/88)


### BREAKING CHANGES

* content is no longer stored in individual buckets but
  inside a common helix-content-bus bucket.
  the content-bus id prefix was modified to not include the `h3` prefix.

## [1.3.18](https://github.com/adobe/helix-content-bus/compare/v1.3.17...v1.3.18) (2021-08-01)


### Bug Fixes

* **deps:** update external fixes ([#96](https://github.com/adobe/helix-content-bus/issues/96)) ([bceb037](https://github.com/adobe/helix-content-bus/commit/bceb037b7056b83257d02b3a9bd17d7434de7ac0))

## [1.3.17](https://github.com/adobe/helix-content-bus/compare/v1.3.16...v1.3.17) (2021-07-28)


### Bug Fixes

* **deps:** update dependency @adobe/helix-status to v9.2.0 ([#93](https://github.com/adobe/helix-content-bus/issues/93)) ([fe603f9](https://github.com/adobe/helix-content-bus/commit/fe603f9ff226c26a89fd721311e3a20e7d2675d9))

## [1.3.16](https://github.com/adobe/helix-content-bus/compare/v1.3.15...v1.3.16) (2021-07-28)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared-config to v1.1.11 ([#92](https://github.com/adobe/helix-content-bus/issues/92)) ([576cb98](https://github.com/adobe/helix-content-bus/commit/576cb98ea136cd833a98705e1048dbb04697b37d))

## [1.3.15](https://github.com/adobe/helix-content-bus/compare/v1.3.14...v1.3.15) (2021-07-27)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared-config to v1.1.10 ([#91](https://github.com/adobe/helix-content-bus/issues/91)) ([0f4fff2](https://github.com/adobe/helix-content-bus/commit/0f4fff2e6f94176e201d08722b65c439054fc296))

## [1.3.14](https://github.com/adobe/helix-content-bus/compare/v1.3.13...v1.3.14) (2021-07-24)


### Bug Fixes

* **deps:** update external fixes ([#90](https://github.com/adobe/helix-content-bus/issues/90)) ([8c2c9fe](https://github.com/adobe/helix-content-bus/commit/8c2c9fe9b5ff94f7d35bd2354944cd06d41ad4cc))

## [1.3.13](https://github.com/adobe/helix-content-bus/compare/v1.3.12...v1.3.13) (2021-07-18)


### Bug Fixes

* **deps:** update external fixes ([#85](https://github.com/adobe/helix-content-bus/issues/85)) ([b382dc2](https://github.com/adobe/helix-content-bus/commit/b382dc2c25c1beda9d1a1e0e86aead0bfe6a6a5b))

## [1.3.12](https://github.com/adobe/helix-content-bus/compare/v1.3.11...v1.3.12) (2021-07-14)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared-body-data to v1.0.11 ([#84](https://github.com/adobe/helix-content-bus/issues/84)) ([ef07096](https://github.com/adobe/helix-content-bus/commit/ef0709611622cc19b8e328c066731c23abad95ac))

## [1.3.11](https://github.com/adobe/helix-content-bus/compare/v1.3.10...v1.3.11) (2021-07-13)


### Bug Fixes

* **deps:** update dependency @adobe/helix-universal to v1.6.5 ([#83](https://github.com/adobe/helix-content-bus/issues/83)) ([370c0ea](https://github.com/adobe/helix-content-bus/commit/370c0ea9f59f3d2b9ca6e740f5102e2063068987))

## [1.3.10](https://github.com/adobe/helix-content-bus/compare/v1.3.9...v1.3.10) (2021-07-13)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared-config to v1.1.8 ([#82](https://github.com/adobe/helix-content-bus/issues/82)) ([1104ab8](https://github.com/adobe/helix-content-bus/commit/1104ab874e0db52d659aa5dec71a0c23e33215fa))

## [1.3.9](https://github.com/adobe/helix-content-bus/compare/v1.3.8...v1.3.9) (2021-07-11)


### Bug Fixes

* **deps:** update external fixes ([#81](https://github.com/adobe/helix-content-bus/issues/81)) ([5463996](https://github.com/adobe/helix-content-bus/commit/546399686de1746caea9b0f9b38adccb6ada131b))

## [1.3.8](https://github.com/adobe/helix-content-bus/compare/v1.3.7...v1.3.8) (2021-07-08)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared-body-data to v1.0.10 ([#80](https://github.com/adobe/helix-content-bus/issues/80)) ([eaa1dd5](https://github.com/adobe/helix-content-bus/commit/eaa1dd5d1fa8d77d45deade46c2f419a8aa57710))

## [1.3.7](https://github.com/adobe/helix-content-bus/compare/v1.3.6...v1.3.7) (2021-07-07)


### Bug Fixes

* **deps:** update dependency @adobe/helix-universal to v1.6.4 ([#79](https://github.com/adobe/helix-content-bus/issues/79)) ([e486626](https://github.com/adobe/helix-content-bus/commit/e486626601a6eb33564ac3e1aaacddf5c3f797d8))

## [1.3.6](https://github.com/adobe/helix-content-bus/compare/v1.3.5...v1.3.6) (2021-07-05)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared-config to v1.1.7 ([#78](https://github.com/adobe/helix-content-bus/issues/78)) ([650e449](https://github.com/adobe/helix-content-bus/commit/650e449e304ef68193d3fb684fc9c2e2f04704ff))

## [1.3.5](https://github.com/adobe/helix-content-bus/compare/v1.3.4...v1.3.5) (2021-07-05)


### Bug Fixes

* **deps:** update external fixes ([#74](https://github.com/adobe/helix-content-bus/issues/74)) ([b4f246b](https://github.com/adobe/helix-content-bus/commit/b4f246b925b975af3bde579cebb8cfbef5e0e9d9))

## [1.3.4](https://github.com/adobe/helix-content-bus/compare/v1.3.3...v1.3.4) (2021-06-17)


### Bug Fixes

* tag value should be escaped ([#73](https://github.com/adobe/helix-content-bus/issues/73)) ([9ff130c](https://github.com/adobe/helix-content-bus/commit/9ff130c23d76415124e9b2dd1098be696bf3fec0))

## [1.3.3](https://github.com/adobe/helix-content-bus/compare/v1.3.2...v1.3.3) (2021-06-14)


### Bug Fixes

* use latest @adobe/helix-fetch and timeoutSignal.clear ([24fe489](https://github.com/adobe/helix-content-bus/commit/24fe4897a145e84896a733cab5ce136ffed6a842))

## [1.3.2](https://github.com/adobe/helix-content-bus/compare/v1.3.1...v1.3.2) (2021-06-13)


### Bug Fixes


test/utils.js
/*
 * Copyright 2021 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

'use strict';

const path = require('path');
const querystring = require('querystring');
const NodeHttpAdapter = require('@pollyjs/adapter-node-http');
const FSPersister = require('@pollyjs/persister-fs');
const { setupMocha } = require('@pollyjs/core');
const { Request } = require('@adobe/helix-fetch');

function setupPolly(opts) {
  setupMocha({
    logging: false,
    recordFailedRequests: true,
    recordIfMissing: false,
    matchRequestsBy: {
      headers: {
        exclude: ['authorization', 'accept-encoding', 'user-agent', 'accept', 'connection', 'x-request-id'],
      },
    },
    adapters: [NodeHttpAdapter],
    persister: FSPersister,
    persisterOptions: {
      fs: {
        recordingsDir: path.resolve(__dirname, 'fixtures'),
      },
    },
    ...opts,
  });
}

function retrofit(fn) {
  const resolver = {
    createURL({ package: pkg, name, version }) {
      return new URL(`https://adobeioruntime.net/api/v1/web/helix/${pkg}/${name}@${version}`);
    },
  };
  return async (params = {}, env = {}, post = false) => {
    const req = post
      ? new Request('https://helix-service.com/publish', {
        method: 'POST',
        body: params,
      })
      : new Request(`https://helix-service.com/publish?${querystring.encode(params)}`);
    const context = {
      resolver,
      env,
      // eslint-disable-next-line no-underscore-dangle
      log: params.__ow_logger,
    };
    const resp = await fn(req, context);
    return {
      statusCode: resp.status,
      body: await resp.text(),
      headers: resp.headers.plain(),
    };
  };
}

module.exports = { setupPolly, retrofit };


test/dev/server.js
/*
 * Copyright 2021 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */
const { DevelopmentServer } = require('@adobe/helix-deploy');
const { main } = require('../../src/index.js');

async function run() {
  const devServer = await new DevelopmentServer(main).init();
  await devServer.start();
}

run().then(process.stdout).catch(process.stderr);


test/index.test.js
/*
 * Copyright 2019 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

/* eslint-env mocha */
/* eslint-disable class-methods-use-this, no-param-reassign */

'use strict';

process.env.HELIX_FETCH_FORCE_HTTP1 = 'true';

const assert = require('assert');

const { condit } = require('@adobe/helix-testutils');

const { main } = require('../src/index.js');
const { setupPolly, retrofit } = require('./utils.js');

// require('dotenv').config();

const index = retrofit(main);

describe('Index Tests', () => {
  setupPolly({
    recordIfMissing: true,
    matchRequestsBy: {
      method: true,
      headers: false,
      body: false,
      order: false,
      url: {
        protocol: true,
        username: false,
        password: false,
        hostname: true,
        port: false,
        pathname: true,
        query: true,
        hash: true,
      },
    },
  });

  beforeEach(function swallowSensitive() {
    const { server } = this.polly;
    server.any().on('beforePersist', (_, recording) => {
      recording.request.headers = recording.request.headers.filter(({ name }) => name !== 'authorization');
      delete recording.request.postData;
    });
  });

  const awsCredentials = {
    AWS_S3_REGION: 'us-east-1',
    AWS_S3_ACCESS_KEY_ID: 'foo',
    AWS_S3_SECRET_ACCESS_KEY: 'bar',
  };

  it('returns 400 if path is missing', async () => {
    const res = await index({
      owner: 'adobe',
      repo: 'theblog',
      ref: 'master',
    });
    assert.strictEqual(res.statusCode, 400);
    assert.match(res.headers['x-error'], /required/);
  });

  it('returns 400 if fstab is missing', async function test() {
    const { server } = this.polly;
    server
      .get('https://helix-code-bus.s3.us-east-1.amazonaws.com/tripodsan/helix-test/main/fstab.yaml?x-id=GetObject')
      .intercept((_, res) => res.status(404));
    const res = await index({
      owner: 'tripodsan',
      repo: 'helix-test',
      ref: 'main',
      path: '/word2md-unit-tests/homepage.md',
    }, awsCredentials);
    assert.strictEqual(res.statusCode, 400);
    assert.match(res.headers['x-error'], /fstab.yaml not found/);
  }).timeout(5000);

  it('returns 400 if no mountpoint matches path', async function test() {
    const fstab = `
    mountpoints:
      /outside: https://adobe.sharepoint.com/sites/TheBlog/Shared%20Documents/theblog
    `;

    const { server } = this.polly;
    server
      .get('https://helix-code-bus.s3.us-east-1.amazonaws.com/tripodsan/helix-test/main/fstab.yaml?x-id=GetObject')
      .intercept((_, res) => res.status(200).send(fstab));

    const res = await index({
      owner: 'tripodsan',
      repo: 'helix-test',
      ref: 'main',
      path: '/word2md-unit-tests/homepage.md',
      useLastModified: 'false',
    }, awsCredentials);
    assert.strictEqual(res.statusCode, 400);
    assert.match(res.headers['x-error'], /not mounted/);
  }).timeout(5000);

  it('returns 200 without schnickschnack with an existing path', async () => {
    const res = await index({
      owner: 'tripodsan',
      repo: 'helix-test',
      ref: 'main',
      path: '/word2md-unit-tests/homepage.md',
    }, awsCredentials);
    assert.strictEqual(res.statusCode, 200);
    assert.strictEqual(res.body, '');
  }).timeout(10000);

  it('creates marker files if missing', async function test() {
    const { server } = this.polly;
    server
      .get('https://helix-content-bus.s3.us-east-1.amazonaws.com/d4b1d4ea6d84b0229bce7cf6806b0bb3470489ab8205a13f75cfe518fa7/.hlx.json?x-id=GetObject')
      .intercept((_, res) => res.status(404));

    const res = await index({
      owner: 'tripodsan',
      repo: 'helix-test',
      ref: 'main',
      path: '/word2md-unit-tests/homepage.md',
    }, awsCredentials);
    assert.strictEqual(res.statusCode, 200);
    assert.strictEqual(res.body, '');
  }).timeout(10000);

  it('returns 200 when publishing an existing item', async () => {
    const res = await index({
      owner: 'tripodsan',
      repo: 'helix-test',
      ref: 'main',
      path: '/word2md-unit-tests/homepage.md',
      action: 'publish',
    }, awsCredentials, true);
    assert.strictEqual(res.statusCode, 200);
    assert.strictEqual(res.body, '');
  }).timeout(5000);

  it('returns 304 with an existing item that did not change', async () => {
    const res = await index({
      owner: 'tripodsan',
      repo: 'helix-test',
      ref: 'main',
      path: '/word2md-unit-tests/homepage.md',
      prefix: 'preview',
      useLastModified: true,
    }, awsCredentials, true);
    assert.strictEqual(res.statusCode, 304);
  }).timeout(5000);

  it('returns 200 with an item that wasn\'t available in S3', async function test() {
    const { server } = this.polly;
    server
      .head('https://helix-content-bus.s3.us-east-1.amazonaws.com/d4b1d4ea6d84b0229bce7cf6806b0bb3470489ab8205a13f75cfe518fa7/preview/word2md-unit-tests/homepage.md')
      .intercept((_, res) => res.status(404));

    const res = await index({
      owner: 'tripodsan',
      repo: 'helix-test',
      ref: 'main',
      path: '/word2md-unit-tests/homepage.md',
      prefix: 'preview',
      useLastModified: true,
    }, awsCredentials, true);
    assert.strictEqual(res.statusCode, 200);
  }).timeout(5000);

  it('returns 404 with a non-existing path', async () => {
    const res = await index({
      owner: 'tripodsan',
      repo: 'helix-test',
      ref: 'main',
      path: '/non-existing/homepage.md',
    }, awsCredentials, true);
    assert.strictEqual(res.statusCode, 404);
  }).timeout(5000);

  it('returns 400 for a unknown action', async () => {
    const res = await index({
      owner: 'tripodsan',
      repo: 'helix-test',
      ref: 'main',
      path: '/word2md-unit-tests/homepage.md',
      action: 'energize',
    }, awsCredentials);
    assert.strictEqual(res.statusCode, 400);
  }).timeout(5000);

  it('returns 404 when copying a missing item', async () => {
    const res = await index({
      owner: 'tripodsan',
      repo: 'helix-test',
      ref: 'main',
      path: '/nonexisting/homepage.md',
      action: 'publish',
    }, awsCredentials, true);
    assert.strictEqual(res.statusCode, 404);
  }).timeout(5000);
});

describe.skip('Live Index Tests', () => {
  condit('Store theblog', condit.hasenvs(['AWS_S3_REGION', 'AWS_S3_ACCESS_KEY_ID', 'AWS_S3_SECRET_ACCESS_KEY']), async () => {
    const res = await index({
      owner: 'adobe',
      repo: 'theblog',
      ref: 'master',
      path: '/en/publish/2020/11/02/high-tech-companies-can-deliver-successful-cx-with-ml-real-time-data.md',
    }, {
      AWS_S3_REGION: process.env.AWS_S3_REGION,
      AWS_S3_ACCESS_KEY_ID: process.env.AWS_S3_ACCESS_KEY_ID,
      AWS_S3_SECRET_ACCESS_KEY: process.env.AWS_S3_SECRET_ACCESS_KEY,
    });
    assert.strictEqual(res.statusCode, 200);
  }).timeout(20000);
  condit('Store spark-website', condit.hasenvs(['AWS_S3_REGION', 'AWS_S3_ACCESS_KEY_ID', 'AWS_S3_SECRET_ACCESS_KEY']), async () => {
    const res = await index({
      owner: 'adobe',
      repo: 'spark-website',
      ref: 'main',
      path: '/express/create/advertisement/cyber-monday.md',
    }, {
      AWS_S3_REGION: process.env.AWS_S3_REGION,
      AWS_S3_ACCESS_KEY_ID: process.env.AWS_S3_ACCESS_KEY_ID,
      AWS_S3_SECRET_ACCESS_KEY: process.env.AWS_S3_SECRET_ACCESS_KEY,
    });
    assert.strictEqual(res.statusCode, 200);
  }).timeout(20000);
  condit('Publish spark-website', condit.hasenvs(['AWS_S3_REGION', 'AWS_S3_ACCESS_KEY_ID', 'AWS_S3_SECRET_ACCESS_KEY']), async () => {
    const res = await index({
      owner: 'adobe',
      repo: 'spark-website',
      ref: 'main',
      path: '/express/create/advertisement/cyber-monday.m',
      action: 'publish',
    }, {
      AWS_S3_REGION: process.env.AWS_S3_REGION,
      AWS_S3_ACCESS_KEY_ID: process.env.AWS_S3_ACCESS_KEY_ID,
      AWS_S3_SECRET_ACCESS_KEY: process.env.AWS_S3_SECRET_ACCESS_KEY,
    });
    assert.strictEqual(res.statusCode, 500);
  }).timeout(20000);
});


src/utils.js
/*
 * Copyright 2021 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

'use strict';

/* eslint-disable no-param-reassign */
const fetchAPI = require('@adobe/helix-fetch');
const { cleanupHeaderValue } = require('@adobe/helix-shared-utils');
const { Response } = require('@adobe/helix-universal');

const { context, ALPN_HTTP1_1 } = fetchAPI;
const { fetch, timeoutSignal } = process.env.HELIX_FETCH_FORCE_HTTP1
  ? context({
    alpnProtocols: [ALPN_HTTP1_1],
    userAgent: 'helix-fetch', // static user agent for test recordings
  })
  /* istanbul ignore next */
  : fetchAPI;

/**
 * Returns fetch compatible options for the given handler options.
 * @param {object} options Handler options
 * @return {object} fetch options.
 */
function getFetchOptions(options) {
  const fetchopts = {
    headers: {},
    ...options,
  };
  if (options.requestId) {
    fetchopts.headers['x-request-id'] = options.requestId;
  }
  if (options.fetchTimeout) {
    fetchopts.signal = timeoutSignal(options.fetchTimeout);
  }
  if (options.lastModified) {
    fetchopts.headers['if-modified-since'] = options.lastModified;
  }
  delete fetchopts.requestId;
  if (fetchopts.token) {
    fetchopts.headers['x-github-token'] = fetchopts.token;
  }
  delete fetchopts.token;

  // delete all secrets
  Object.keys(fetchopts)
    .forEach((key) => {
      if (key.match(/^[A-Z0-9_]+$/)) {
        delete fetchopts[key];
      }
    });
  return fetchopts;
}

/**
 * Create an error response.
 */
function createErrorResponse({
  e, msg, status, log,
}) {
  const message = (e && e.message) || msg;
  if (log) {
    const args = [message];
    if (e) {
      args.push(e);
    }
    log.error(...args);
  }
  /* istanbul ignore next (default 500 status) */
  return new Response('', {
    status: (e && e.status) || status || 500,
    headers: {
      'x-error': cleanupHeaderValue(message),
    },
  });
}

/**
 * Escape tag value, so it's acceptable for AWS S3.
 *
 * @param value {string} value to escape
 *
 * @returns escaped tag value
 *
 * The allowed characters across services are: letters, numbers, and spaces representable in UTF-8,
 * and the following characters: + - = . _ : / @.
 */
function escapeTagValue(value) {
  let escaped;
  if (value.match(/https:\/\/.*\.sharepoint\.com/)) {
    escaped = decodeURI(value);
  } else if (value.match(/https:\/\/drive.google.com/)) {
    const query = value.indexOf('?');
    escaped = value.substring(0, query !== -1 ? query : value.length);
  } else {
    escaped = value.replace(/[^A-Za-z0-9- +=._:/@]/g, '_');
  }
  return escaped;
}

module.exports = {
  fetch,
  getFetchOptions,
  createErrorResponse,
  escapeTagValue,
};


src/content-proxy.js
/*
 * Copyright 2021 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

'use strict';

const { Response } = require('@adobe/helix-fetch');
const { logLevelForStatusCode, propagateStatusCode } = require('@adobe/helix-shared-utils');
const { fetch, getFetchOptions } = require('./utils.js');

/**
 * Pass through headers that we keep in the forwarded response.
 */
const PASSTHROUGH_HEADERS = [
  'content-type',
  'last-modified',
  'x-source-location',
];

/**
 * Create the URL to fetch content from.
 *
 * @param {object}   opts options
 * @param {string}   opts.owner the GitHub org or username
 * @param {string}   opts.repo the GitHub repository
 * @param {string}   opts.ref the GitHub ref
 * @param {string}   opts.path the path of the file to retrieve
 * @param {string}   opts.mp mountpoint
 * @param {object}   opts.log a Helix-Log instance
 * @param {object}   opts.options Helix Fetch options
 * @param {Resolver} opts.resolver Version lock helper
 *
 * @returns {string} URL to fetch content from
 */
function createURL(opts) {
  const {
    owner, repo, ref, path, mp, options, resolver,
  } = opts;

  const url = resolver.createURL({
    package: 'helix-services',
    name: 'content-proxy',
    version: 'v2',
  });
  url.searchParams.append('owner', owner);
  url.searchParams.append('repo', repo);
  url.searchParams.append('ref', ref);
  url.searchParams.append('path', path);

  if (mp) {
    url.searchParams.append('mpType', mp.type);
    url.searchParams.append('mpRelPath', mp.relPath);
    url.searchParams.append('mpURL', mp.url);
  }
  url.searchParams.append('rid', options.requestId);
  return url.href;
}

/**
 * Fetches a document, either by using the content-proxy service or going to a CDN.
 *
 * @param {object}   opts options
 * @param {string}   opts.owner the GitHub org or username
 * @param {string}   opts.repo the GitHub repository
 * @param {string}   opts.ref the GitHub ref
 * @param {string}   opts.path the path of the file to retrieve
 * @param {string}   opts.mp mountpoint
 * @param {object}   opts.log a Helix-Log instance
 * @param {object}   opts.options Helix Fetch options
 * @param {Resolver} opts.resolver Version lock helper
 *
 * @returns {Response} response
 */
async function contentProxy(opts) {
  const {
    log, options,
  } = opts;

  const url = createURL(opts);
  log.info(`Fetching content from: ${url}`);

  const fetchopts = getFetchOptions(options);

  try {
    const resp = await fetch(url, fetchopts);
    const body = await resp.buffer();
    if (resp.ok) {
      const headers = {};
      PASSTHROUGH_HEADERS.forEach((name) => {
        const value = resp.headers.get(name);
        if (value) {
          headers[name] = value;
        }
      });
      return new Response(body, {
        status: 200,
        headers,
      });
    }
    log[logLevelForStatusCode(resp.status)](`Unable to fetch ${url} (${resp.status}): ${resp.headers.get('x-error')}`);
    return new Response(body, {
      status: propagateStatusCode(resp.status),
      headers: {
        'x-error': resp.headers.get('x-error'),
        'cache-control': 'private, no-cache',
      },
    });
  } finally {
    if (fetchopts.signal) {
      fetchopts.signal.clear();
    }
  }
}

module.exports = {
  contentProxy,
};


src/index.js
/*
 * Copyright 2021 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

'use strict';

const crypto = require('crypto');

const { AbortError, FetchError } = require('@adobe/helix-fetch');
const { logger } = require('@adobe/helix-universal-logger');
const wrap = require('@adobe/helix-shared-wrap');
const bodyData = require('@adobe/helix-shared-body-data');
const { MountConfig } = require('@adobe/helix-shared-config');
const { wrap: helixStatus } = require('@adobe/helix-status');
const { Response } = require('@adobe/helix-universal');

const { contentProxy } = require('./content-proxy.js');
const { AWSStorage } = require('./storage.js');
const { createErrorResponse } = require('./utils.js');

/**
 * Parse a boolean given as either a string or a boolean.
 *
 * @param {any} value value to parse
 * @param {boolean} defaultValue default value to use if value is unset
 * @returns true, false or the default value
 */
function parseBoolean(value, defaultValue) {
  if (value === 'false' || value === Boolean(false)) {
    return false;
  }
  return value ? !!value : defaultValue;
}

/**
 * Fetches content from content-proxy and stores it in an S3 bucket.
 *
 * @param {Request} req request object
 * @param {Context} context request context
 * @returns {Response} the response
 */
async function main(req, context) {
  const { env, log, resolver } = context;
  const {
    AWS_S3_REGION, AWS_S3_ACCESS_KEY_ID, AWS_S3_SECRET_ACCESS_KEY,
    HTTP_TIMEOUT_EXTERNAL,
  } = env;

  const {
    owner, repo, ref, path, prefix = 'live', action = 'update',
  } = context.data;

  const useLastModified = parseBoolean(context.data.useLastModified, false);

  if (!(owner && repo && ref && path)) {
    return createErrorResponse({
      status: 400,
      msg: 'owner, repo, ref, and path parameters are required',
    });
  }

  let codeStorage;
  let fstab;

  try {
    codeStorage = new AWSStorage({
      AWS_S3_REGION,
      AWS_S3_ACCESS_KEY_ID,
      AWS_S3_SECRET_ACCESS_KEY,
      bucket: 'helix-code-bus',
      readOnly: true,
      log,
    });
    const buffer = await codeStorage.load(`${owner}/${repo}/${ref}/fstab.yaml`);
    if (!buffer) {
      return createErrorResponse({
        log,
        status: 400,
        msg: `${owner}/${repo}/${ref}/fstab.yaml not found in bucket 'helix-code-bus'`,
      });
    }
    fstab = await new MountConfig().withSource(buffer.toString()).init();
  } catch (e) {
    /* istanbul ignore next */
    return createErrorResponse({ e, log });
  } finally {
    /* istanbul ignore else */
    if (codeStorage) {
      codeStorage.close();
    }
  }

  const mp = fstab.match(path);
  if (!mp) {
    return createErrorResponse({
      log,
      status: 400,
      msg: `path specified is not mounted in fstab.yaml: ${path}`,
    });
  }

  const options = {
    cache: 'no-store',
    fetchTimeout: HTTP_TIMEOUT_EXTERNAL || 20000,
    requestId: req.headers.get('x-request-id')
    || req.headers.get('x-cdn-request-id')
    || req.headers.get('x-openwhisk-activation-id')
    || '',
    token: req.headers.get('x-github-token'),
  };

  let contentStorage;

  try {
    const sha256 = crypto
      .createHash('sha256')
      .update(mp.url)
      .digest('hex');
    const contentBusId = `${sha256.substr(0, 59)}`;

    contentStorage = new AWSStorage({
      AWS_S3_REGION,
      AWS_S3_ACCESS_KEY_ID,
      AWS_S3_SECRET_ACCESS_KEY,
      bucket: 'helix-content-bus',
      log,
    });

    if (action === 'update') {
      const key = `${contentBusId}/${prefix}${path}`;

      if (useLastModified) {
        const metadata = await contentStorage.metadata(key);
        if (metadata) {
          options.lastModified = metadata['x-source-last-modified'];
        }
      }
      const res = await contentProxy({
        owner, repo, ref, path, mp, log, options, resolver,
      });
      if (!res.ok) {
        return res;
      }
      await contentStorage.store(key, res);

      // set container info if not present
      const infoKey = `${contentBusId}/.hlx.json`;
      if (!await contentStorage.load(infoKey)) {
        const meta = {
          mountpoint: mp.url,
          'original-repository': `${owner}/${repo}`,
        };
        await contentStorage.storeData(infoKey, Buffer.from(JSON.stringify(meta, null, 2)), 'application/json', meta);

        // create marker file so that the root folder is easy to find.
        await contentStorage.storeData(`${contentBusId}--${owner}--${repo}`, '', 'text/plain', meta);
      }
      return new Response('', { status: 200 });
    }

    if (action === 'publish') {
      await contentStorage.copy(`${contentBusId}/preview${path}`, `${contentBusId}/live${path}`);
      return new Response('', { status: 200 });
    }

    return createErrorResponse({
      status: 400,
      msg: `Action unknown: ${action}`,
    });
  } catch (e) {
    /* istanbul ignore next */
    if (e instanceof AbortError) {
      return new Response(e.message, {
        status: 504,
      });
    }
    /* istanbul ignore next */
    if (e instanceof FetchError) {
      if (e.code === 'ECONNRESET') {
        // connection reset by host: temporary network issue
        return new Response(e.message, {
          status: 504,
        });
      }
    }
    /* istanbul ignore next */
    return createErrorResponse({ e, log });
  } finally {
    /* istanbul ignore else */
    if (contentStorage) {
      contentStorage.close();
    }
  }
}

module.exports.main = wrap(main)
  .with(bodyData)
  .with(helixStatus)
  .with(logger.trace)
  .with(logger);
